testsuite/tests/typing-misc/constraints.ml

```diff
@@ -85,17 +85,31 @@ Error: The definition of abs contains a cycle:
        'a is_an_object as 'a
 |}];;
 
-module PR6505a = struct
+module PR6505a_old = struct
   type 'o is_an_object = < .. > as 'o
   and ('k,'l) abs = 'l constraint 'k = 'l is_an_object
   let y : ('o, 'o) abs = object end
 end;;
+[%%expect{|
+Line 3, characters 7-9:
+3 |   and ('k,'l) abs = 'l constraint 'k = 'l is_an_object
+           ^^
+Error: Constraints are not satisfied in this type.
+       Type 'l is_an_object should be an instance of < .. > is_an_object
+|}]
+
+module PR6505a = struct
+  type 'o is_an_object = < .. > as 'o
+  and ('k,'l) abs = 'l constraint 'k = 'l is_an_object constraint 'l = < .. >
```

Comment:
I'm a bit confused about what is the status of this testsuite example. My understanding is that the original code

```
type ('k, 'l) abs = 'l constraint 'k = 'l is_an_object
```

was written with the understanding that `constraint 'k = 'l is_an_object` would enforce that `'k` be an object type, and also make it equal to `'l`. Now the checker is more strict and this constraint is rejected unless we also force `'l` to be an object type in a separate constraint -- so the constraint `is_an_object` does not seem to be of any use anymore, unless I am missing something. Assuming we are trying to write a short example of interesting constraint (and not just to record the fix to an existing example), should this not be written

```ocaml
type ('k, 'l) abs = 'l constraint 'k = 'l constraint 'l = < .. >
```

? If I understand correctly, this is exactly equivalent, as we cannot use `'o is_an_object` to imply an additional constraint anymore.

Reply:
Good point.
Since the problem is not so much the module, but rather that `y` should not have an ill-founded type, I have changed the example to
```ocaml
  type 'o is_an_object = < .. > as 'o
  type ('k,'l) abs = 'l constraint 'k = 'l is_an_object
```
which also answers your other question.